File: app/src/androidTest/java/org/kiwix/kiwixmobile/NetworkTest.java
/*
 * Kiwix Android
 * Copyright (C) 2018  Kiwix <android.kiwix.org>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package org.kiwix.kiwixmobile;

import android.Manifest;
import android.util.Log;
import androidx.test.espresso.DataInteraction;
import androidx.test.espresso.IdlingPolicies;
import androidx.test.espresso.IdlingRegistry;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.rule.ActivityTestRule;
import androidx.test.rule.GrantPermissionRule;
import com.schibsted.spain.barista.interaction.BaristaMenuClickInteractions;
import com.schibsted.spain.barista.interaction.BaristaSleepInteractions;
import java.io.InputStream;
import java.util.concurrent.TimeUnit;
import org.junit.After;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Ignore;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.kiwix.kiwixmobile.data.ZimContentProvider;
import org.kiwix.kiwixmobile.di.components.DaggerTestComponent;
import org.kiwix.kiwixmobile.di.components.TestComponent;
import org.kiwix.kiwixmobile.main.MainActivity;
import org.kiwix.kiwixmobile.testutils.TestUtils;
import org.kiwix.kiwixmobile.utils.KiwixIdlingResource;

import static androidx.test.InstrumentationRegistry.getInstrumentation;
import static androidx.test.espresso.Espresso.onData;
import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.action.ViewActions.longClick;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static com.schibsted.spain.barista.interaction.BaristaClickInteractions.clickOn;
import static com.schibsted.spain.barista.interaction.BaristaDialogInteractions.clickDialogPositiveButton;
import static com.schibsted.spain.barista.interaction.BaristaMenuClickInteractions.clickMenu;
import static com.schibsted.spain.barista.interaction.BaristaSwipeRefreshInteractions.refresh;
import static org.hamcrest.CoreMatchers.allOf;
import static org.kiwix.kiwixmobile.testutils.TestUtils.TEST_PAUSE_MS;
import static org.kiwix.kiwixmobile.testutils.TestUtils.getResourceString;
import static org.kiwix.kiwixmobile.testutils.TestUtils.withContent;

/**
 * Created by mhutti1 on 14/04/17.
 */
@RunWith(AndroidJUnit4.class)
public class NetworkTest {
  private static final String NETWORK_TEST_TAG = "KiwixNetworkTest";
  @Rule
  public ActivityTestRule<MainActivity> mActivityTestRule = new ActivityTestRule<>(
      MainActivity.class, false, false);
  @Rule
  public GrantPermissionRule readPermissionRule =
      GrantPermissionRule.grant(Manifest.permission.READ_EXTERNAL_STORAGE);
  @Rule
  public GrantPermissionRule writePermissionRule =
      GrantPermissionRule.grant(Manifest.permission.WRITE_EXTERNAL_STORAGE);
  //@Inject
  //MockWebServer mockWebServer;

  @BeforeClass
  public static void beforeClass() {
    IdlingPolicies.setMasterPolicyTimeout(180, TimeUnit.SECONDS);
    IdlingPolicies.setIdlingResourceTimeout(180, TimeUnit.SECONDS);
    IdlingRegistry.getInstance().register(KiwixIdlingResource.getInstance());
  }

  @Before
  public void setUp() {

    TestComponent component = DaggerTestComponent.builder().context(
        getInstrumentation().getTargetContext().getApplicationContext()).build();

    KiwixApplication.setApplicationComponent(component);

    new ZimContentProvider().setupDagger();
    component.inject(this);
    InputStream library = NetworkTest.class.getClassLoader().getResourceAsStream("library.xml");
    InputStream metalinks =
        NetworkTest.class.getClassLoader().getResourceAsStream("test.zim.meta4");
    InputStream testzim = NetworkTest.class.getClassLoader().getResourceAsStream("testzim.zim");
    //try {
    //  byte[] libraryBytes = IOUtils.toByteArray(library);
    //  mockWebServer.enqueue(new MockResponse().setBody(new String(libraryBytes)));
    //  byte[] metalinkBytes = IOUtils.toByteArray(metalinks);
    //  mockWebServer.enqueue(new MockResponse().setBody(new String(metalinkBytes)));
    //  mockWebServer.enqueue(new MockResponse().setHeader("Content-Length", 357269));
    //  Buffer buffer = new Buffer();
    //  buffer.write(IOUtils.toByteArray(testzim));
    //  buffer.close();
    //  mockWebServer.enqueue(new MockResponse().setBody(buffer));
    //} catch (IOException e) {
    //  e.printStackTrace();
    //}
  }

  @Test
  @Ignore("Broken in 2.5")//TODO: Fix in 3.0
  public void networkTest() {

    mActivityTestRule.launchActivity(null);

    BaristaSleepInteractions.sleep(TEST_PAUSE_MS);
    BaristaMenuClickInteractions.clickMenu(getResourceString(R.string.menu_zim_manager));

    TestUtils.allowPermissionsIfNeeded();

    onData(withContent("wikipedia_ab_all_2017-03")).inAdapterView(withId(R.id.libraryList)).perform(click());

    try {
      onView(withId(android.R.id.button1)).perform(click());
    } catch (RuntimeException e) {
    }

    clickOn(R.string.local_zims);

    try {
      onData(allOf(withId(R.id.zim_swiperefresh)));
      refresh(R.id.zim_swiperefresh);
      Thread.sleep(500);
    } catch (InterruptedException e) {
      e.printStackTrace();
    }

    // Commented out the following which assumes only 1 match - not always safe to assume as there may
    // already be a similar file on the device.
    // onData(withContent("wikipedia_ab_all_2017-03")).inAdapterView(withId(R.id.zimfilelist)).perform(click());

    // Find matching zim files on the device
    try {
      DataInteraction dataInteraction =
          onData(withContent("wikipedia_ab_all_2017-03")).inAdapterView(withId(R.id.zimfilelist));
      // TODO how can we get a count of the items matching the dataInteraction?
      dataInteraction.atPosition(0).perform(click());

      clickMenu(R.string.menu_zim_manager);

      DataInteraction dataInteraction1 =
          onData(withContent("wikipedia_ab_all_2017-03")).inAdapterView(withId(R.id.zimfilelist));
      dataInteraction1.atPosition(0).perform(longClick()); // to delete the zim file
      clickDialogPositiveButton();
    } catch (Exception e) {
      Log.w(NETWORK_TEST_TAG, "failed to interact with local ZIM file: " + e.getLocalizedMessage());
    }
  }

  @After
  public void finish() {
    IdlingRegistry.getInstance().unregister(KiwixIdlingResource.getInstance());
  }
}

File: app/src/androidTest/java/org/kiwix/kiwixmobile/NetworkTest.kt
/*
 * Kiwix Android
 * Copyright (c) 2019 Kiwix <android.kiwix.org>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 */
package org.kiwix.kiwixmobile

import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.filters.SmallTest
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.test.runTest
import kotlinx.coroutines.withTimeout
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import okhttp3.logging.HttpLoggingInterceptor.Level.BASIC
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import okhttp3.mockwebserver.SocketPolicy
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.kiwix.kiwixmobile.core.data.remote.KiwixService
import org.kiwix.kiwixmobile.core.utils.files.Log
import org.kiwix.kiwixmobile.testutils.RetryRule
import org.kiwix.sharedFunctions.TEST_PORT
import java.net.InetAddress
import java.util.concurrent.TimeUnit.SECONDS

/**
 * Created by mhutti1 on 14/04/17.
 */
@SmallTest
@RunWith(AndroidJUnit4::class)
class NetworkTest {
  @Rule
  @JvmField
  val retryRule = RetryRule()

  private lateinit var mockWebServer: MockWebServer
  private lateinit var kiwixService: KiwixService

  @Before
  fun setUp() {
    mockWebServer = MockWebServer()
    mockWebServer.start(InetAddress.getByName("127.0.0.1"), TEST_PORT)
    kiwixService = KiwixService.ServiceCreator.newHackListService(
      OkHttpClient().newBuilder()
        .connectTimeout(TEST_TIMEOUT, SECONDS)
        .readTimeout(TEST_TIMEOUT, SECONDS)
        .callTimeout(TEST_TIMEOUT, SECONDS)
        .addNetworkInterceptor(HttpLoggingInterceptor().apply { level = BASIC })
        .build(),
      mockWebServer.url("/").toString()
    )
    Log.d(TAG, "MockWebServer started on port $TEST_PORT")
  }

  @After
  fun tearDown() {
    mockWebServer.shutdown()
    Log.d(TAG, "MockWebServer shut down")
  }

  private fun getResourceAsString(name: String): String =
    javaClass.classLoader!!.getResourceAsStream(name)!!.bufferedReader().readText()

  @Test
  fun testNetworkSuccess() = runTest {
    val libraryXml = getResourceAsString("library.xml")
    mockWebServer.enqueue(
      MockResponse()
        .setResponseCode(200)
        .setBody(libraryXml)
    )

    val response = kiwixService.getLibraryPage(mockWebServer.url("/v2/entries").toString())
    val recordedRequest = mockWebServer.takeRequest()
    Log.d(TAG, "testNetworkSuccess: code=${response.code()} bodyLength=${response.body()?.length}")
    Log.d(TAG, "testNetworkSuccess: method=${recordedRequest.method} path=${recordedRequest.path}")
    Log.d(
      TAG,
      "testNetworkSuccess: code=${response.code()} bodyLength=${response.body()?.length} bodyPreview=${
        response.body()?.take(100)
      }"
    )

    assertEquals("GET", recordedRequest.method)
    assertTrue(recordedRequest.path!!.startsWith("/v2/entries"))
    assertTrue(response.isSuccessful)
    assertEquals(200, response.code())
    assertEquals(libraryXml, response.body())
  }

  @Test
  fun testNetworkError() = runTest {
    mockWebServer.enqueue(MockResponse().setResponseCode(500))

    val response = kiwixService.getLibraryPage(mockWebServer.url("/v2/entries").toString())
    Log.d(TAG, "testNetworkError: code=${response.code()} body=${response.body()}")

    assertFalse(response.isSuccessful)
    assertEquals(500, response.code())
    assertNull(response.body())
    assertNotNull(response.errorBody())
  }

  @Test
  fun testNetworkNotFound() = runTest {
    mockWebServer.enqueue(MockResponse().setResponseCode(404))

    val response = kiwixService.getLibraryPage(mockWebServer.url("/v2/entries").toString())
    Log.d(TAG, "testNetworkNotFound: code=${response.code()} body=${response.body()}")

    assertFalse(response.isSuccessful)
    assertEquals(404, response.code())
    assertNull(response.body())
    assertNotNull(response.errorBody())
  }

  @Test
  fun testNetworkEmptyResponse() = runTest {
    mockWebServer.enqueue(
      MockResponse()
        .setResponseCode(200)
        .setBody("")
    )

    val response = kiwixService.getLibraryPage(mockWebServer.url("/v2/entries").toString())
    Log.d(TAG, "testNetworkEmptyResponse: code=${response.code()} body='${response.body()}'")

    assertTrue(response.isSuccessful)
    assertEquals(200, response.code())
    assertEquals("", response.body())
  }

  @Test
  fun testNetworkTimeout() {
    mockWebServer.enqueue(
      MockResponse().apply { socketPolicy = SocketPolicy.NO_RESPONSE }
    )

    var exceptionThrown = false
    try {
      runBlocking {
        withTimeout((TEST_TIMEOUT + 2L) * 1000L) {
          kiwixService.getLibraryPage(mockWebServer.url("/v2/entries").toString())
        }
      }
    } catch (e: Exception) {
      Log.e(TAG, "testNetworkTimeout: caught expected ${e::class.java.simpleName}: ${e.message}")
      exceptionThrown = true
    }

    assertTrue("Expected a timeout exception to be thrown", exceptionThrown)
  }

  companion object {
    private const val TAG = "NetworkTest"
    private const val TEST_TIMEOUT = 5L
  }
}
